Feat: Added Command Console

Sahil-Chhoker opened the pull request:
### Summary
Adds a command console in mesa as per #2683.
Closes #2683.

### Implementation
This feature consists of four main classes:
1. `ConsoleEntry`: Defines how the data is entered into the console.
2. `CaptureOutput`: A context manager for capturing stdout and stderr output.
3. `InteractiveConsole`: A custom interactive Python console with output capturing capabilities. This class extends `code.InteractiveConsole` for a console like behaviour.
4. `ConsoleManager`: A console manager for executing Python code interactively. This class provides functionality to execute Python code in an interactive console environment, maintain command history, and handle multi-line code blocks.

For more information refer to the docstrings in the file.

### Usage Examples
```python
from mesa.visualization import CommandConsole
from .agents import myAgent
import numpy as np
import math

console_kwargs = {
    "additional_imports": { # additional imports to be used in command console
        "myAgent": myAgent,
        "np": np,
        "math": math,
    }
}

page = SolaraViz(
    model,
    components=[CommandConsole], # Pass the CommandConsole as a component
    model_params=model_params,
    name="Test Model",
    **console_kwargs
)
page
```

### How it looks:
![image](https://github.com/user-attachments/assets/b3ed7158-7868-44be-be4b-46e8305e79b4)

Sahil-Chhoker commented:
I would love to hear your thoughts on this, @quaquel.

quaquel commented:
I think this is really nice. 

Unfortunately, I don't have time to play around with it at the moment. However, I know you have been testing this extensively. I'll try to review soon, but I have various other obligations at the moment.

quaquel commented on mesa/visualization/solara_viz.py:
```diff
@@ -60,6 +61,7 @@ def SolaraViz(
     model_params=None,
     name: str | None = None,
     use_threads: bool = False,
+    additional_imports: dict | None = None,
```
why did you do this? Is it not possible to import stuff you need in the console yourself?

Also, perhaps call this console_kwargs or something because we might use it later for additional stuff?

Sahil-Chhoker replied:
I tried, but its not possible to import custom made agent classes into the console environment.

quaquel reviewed (commented):
Code looks clear and well documented. Great stuff!

I still want to try this myself before approving. Hopefully later today if all goes well.

quaquel commented:
Started playing with it. I think it is great. Below is a WIP list of things I run into. Not all of this needs to be addressed in this PR.

1. Is it possible to "inherit" all imports from `app.py`? That is, populate the namespace of the console based on what is defined in the file from which the solara front end is started.
2. Great to have `model` readily available. This makes dynamically exploring the model easy. 
3. It would make sense if the view jumps to the bottom after each command. Currently, I have scroll down every time.
4. ...

Sahil-Chhoker commented:
> 1. Is it possible to "inherit" all imports from `app.py`? That is, populate the namespace of the console based on what is defined in the file from which the solara front end is started.

Don't know yet, it doesn't seem very plausible because those are the imports we can't see. Will try to find a way.
 
> 2. Great to have `model` readily available. This makes dynamically exploring the model easy.

Glad to know.

> 3. It would make sense if the view jumps to the bottom after each command. Currently, I have to scroll down every time.

This feature makes sense, I will try to incorporate it.

> 4. Is it possible to "force" and update of the display without calling `model.step()`? For example, If I update the wealth of an agent in the boltzmann wealth model, it would be nice if I see the color of an agent changing.

I think I've added that already in the recent commits, I also realized that was very annoying, that's why I added `force_update()` after every execution. Can you please check if that line is there in your code?

> 5. Currently, there are two windows: the console history and the line where you type new commands. Most consoles in IDE's integrate this into a single window where the last line is just where you type commands. Is something like this possible with this code base as well at some point in the future?

Most likely possible, but I will have to go through the solara documentation thoroughly for that (because just adding the enter took long enough).

> 6. Discussion point: where to put the console? Currently, it is on the left-hand side, but this is getting quite crowded. I like the convention in most IDE's to have a console at the bottom right. How difficult would it be to do this at some point?

same as the answer in the 5th. Plus I would love to get this PR merged and look back at the code when implementing the variable explorer, because that will also take up some space.

quaquel commented:
Thanks for the responses. None of these comments are blocking approval. They are merely a wishlist from my side. 

1. Is what can be seen based on where you create the console?
3. This would be the main thing to add if it is easy before merging
4. It is working, I just missed it. I do notice for boltzman that gini is not updated. I understand why given how all this is implemented, but for a future extension this would be nice to have (probably requires using some of the experimental signal stuff)
5. Ok
6. Ok
7.  New wish: in most consoles, arrow key up gives you the last command. Is something like this available in `code` so that we can add support for the history of commands in the future?

Sahil-Chhoker commented:
1. The model instance that is to be used is passed onto the `SolaraViz`, that can be directly used in the console environment but that is not the case with other imports in the `app.py`. There may be some kind of way there, if I search hard enough I may find it, but not soon.
2. .
3. I am working on it now. Just found a source after searching a bit. Will see if be able to do it in the next 2 hours, otherwise it will be postponed until tomorrow.
4. Ok
5. .
6. .
7. I also want that, but its not easy to make custom key handlers that's the main reason it is postponed until now. Will try to ask some friends that are more knowledge in the `javascript` web development field. That may act as reference. (but also for me its a must have)

Sahil-Chhoker commented:
Auto-scrolling is implemented, simple fix I found was to use `ChatBox` from `solara.lab` which already contains this functionality.

Corvince commented:
Thank you very much @quaquel for starting the review. I, too, wanted to try it out before reviewing this PR. But I still haven't found the time to dedust my local mesa workspace. 

> Thanks for the responses. None of these comments are blocking approval. They are merely a wishlist from my side.

This should be quite important. I think it would be good to take those (very good) ideas to a follow-up PR and try to merge this soonish. Its an awesome PR, I can already tell without doing a proper review right now.

quaquel reviewed (commented):
I am reviewing the code atm. One request: can you add the console stuff to `/docs/apis/visualization.md`? I am wondering if some of the docstrings will render correctly.

Sahil-Chhoker commented:
> One request: can you add the console stuff to `/docs/apis/visualization.md`? I am wondering if some of the docstrings will render correctly.

I don't know how the doc stuff works, should I add this?
```python
\## Command Console

\```{eval-rst}
\.. automodule:: mesa.visualization.command_console
   :members:
   :undoc-members:
   :show-inheritance:
\```

```
ignore the '\'

quaquel commented on mesa/visualization/command_console.py:
```diff
@@ -0,0 +1,408 @@
+        console (InteractiveConsole): Python's interactive console instance
+        buffer (list): Buffer for storing multi-line code blocks
+        history (list[ConsoleEntry]): List of console entries containing commands and their outputs
+    Methods:
+        execute_code(code_line: str, set_input_text: Callable[[str], None]) -> None:
+            Executes the provided code line and updates console history
+        clear_console() -> None:
+            Clears the console history and resets console state
+        get_entries() -> list[ConsoleEntry]:
+            Returns the list of console entries
```
```suggestion
```

there is no need to document the methods in the class level docstring. If done this way, they are rendered twice on read the docs

quaquel commented on mesa/visualization/command_console.py:
```diff
@@ -0,0 +1,408 @@
+            Clears the console history and resets console state
+        get_entries() -> list[ConsoleEntry]:
+            Returns the list of console entries
+    Special Commands:
+        history: Shows the command history
+        cls: Clears the console screen
+        tips: Shows available console commands and usage tips
```
these don't render as a bulleted list because special commands is not an official google docstring command as far as I know.

Sahil-Chhoker replied:
Special Commands:
        1. history: Shows the command history
        2. cls: Clears the console screen
        3. tips: Shows available console commands and usage tips

quaquel replied:
you can try and see. My workflow is to just built the docs (can be seen if you open up 'all checks have passed', third last item). You can view the docs by clicking on the succesfull build.

quaquel reviewed (approved):
2 minor doc comments, but otherwise good to go.

Amazing work!

quaquel commented on mesa/visualization/command_console.py:
```diff
@@ -0,0 +1,401 @@
+This module provides a set of classes and functions to create an interactive Python console
+that can be embedded in a web browser. It supports command history, multi-line code blocks,
+and special commands for console management.
+Classes:
+    ConsoleEntry: Stores command console entries with their outputs and states.
+    CaptureOutput: Context manager for capturing stdout and stderr output.
+    InteractiveConsole: Custom interactive Python console with output capturing.
+    ConsoleManager: Manages the execution of Python code and console state.
+Functions:
+    format_command_html: Formats console commands as HTML for display.
+    format_output_html: Formats console output as HTML for display.
+    CommandConsole: Solara component for the interactive console interface.
```
```suggestion
```

I missed this one, it can also be fully removed because its all documented below.

Sahil-Chhoker commented:
Yeah docs look good now, its good to go!

quaquel commented:
It would be good to open an overview issue with a wishlist of possible extensions based on the ideas discussed here.

EwoutH commented:
Congratulations on getting this PR merged! Awesome feature already and really excited in what it can involve into!
